watcher/watcher.go

```diff
@@ -600,75 +588,208 @@ func (w *Watcher) checkState(
+			unbalancedState = models.StateOverdimensioned
+		}
+
+		// If usage is above limit, scale up.
+		if isDown == false && 100*roomCount.Occupied >= roomCount.Total()*trigger.Limit {
+			w.Logger.Debug("Usage is above limit. Should scale up")
+			inSync = false
 			changedState = true
-		} else if nowTimestamp-scheduler.LastScaleOpAt > int64(autoScalingInfo.Up.Cooldown) {
-			shouldScaleUp = true
+			shouldScaleUpOrDown = true
+			scheduler.State = unbalancedState
+			scheduler.StateLastChangedAt = nowTimestamp
+			delta, err := dynamicDelta(trigger, roomCount)
+			if err != nil {
+				return 0, false, false, true, err
+			}
+			scaleInfo.SendUsage(
+				w.SchedulerName,
+				trigger.Metric, point, roomCount.Total(),
+			)
+			return delta, shouldScaleUpOrDown, changedState, inSync, nil
+		}
+
+		isAboveThreshold, err := scaleInfo.SendUsageAndReturnStatus(
+			w.SchedulerName,
+			trigger.Metric,
+			capacity, point, roomCount.Total(),
+			threshold, usage,
+		)
+		if err != nil {
+			return 0, shouldScaleUpOrDown, changedState, inSync, err
+		}
+
+		if isAboveThreshold {
+			inSync = false
+			if scheduler.State != unbalancedState {
+				changedState = true
+			} else if nowTimestamp-scheduler.LastScaleOpAt > int64(scalingPolicy.Cooldown) {
+				shouldScaleUpOrDown = true
+			}
+		}
+
+		if shouldScaleUpOrDown == true {
+			delta, err = dynamicDelta(trigger, roomCount)
+			if err != nil {
+				return 0, false, false, true, err
+			}
+			if isDown {
+				delta = -delta
+			}
+			break
 		}
 	}
 
-	threshold = autoScalingInfo.Down.Trigger.Threshold
-	usage = float32(autoScalingInfo.Down.Trigger.Usage) / 100
-	capacity = autoScalingInfo.Down.Trigger.Time / w.AutoScalingPeriod
+	if changedState == true {
+		scheduler.State = unbalancedState
+		scheduler.StateLastChangedAt = nowTimestamp
+	}
+
+	return delta, shouldScaleUpOrDown, changedState, inSync, nil
+}
 
-	isAboveThreshold, err = w.ScaleDownInfo.SendUsageAndReturnStatus(
+func (w *Watcher) checkLegacyTrigger(
+	autoScalingInfo *models.AutoScaling,
+	scalingPolicy *models.ScalingPolicy,
+	scaleInfo *models.ScaleInfo,
+	roomCount *models.RoomsStatusCount,
+	scheduler *models.Scheduler,
+	nowTimestamp int64,
+) (int, bool, bool, bool, error) { // delta, shouldScaleUpOrDown, changedState, inSync
+	isDown := autoScalingInfo.Down == scalingPolicy
+	shouldScaleUpOrDown := false
+	changedState := false
+	inSync := true
+
+	threshold := scalingPolicy.Trigger.Threshold
+	usage := float32(scalingPolicy.Trigger.Usage) / 100
+	capacity := scalingPolicy.Trigger.Time / w.AutoScalingPeriod
+	point := roomCount.Occupied
+	unbalancedState := models.StateSubdimensioned
+
+	if isDown {
+		usage = 1 - usage
+		point = roomCount.Ready
+		unbalancedState = models.StateOverdimensioned
+	}
+
+	// If usage is above limit, scale up.
+	if 100*roomCount.Occupied >= roomCount.Total()*autoScalingInfo.Up.Trigger.Limit {
+		scaleInfo.SendUsage(
+			w.SchedulerName,
+			"legacy", point, roomCount.Total(),
```

Comment:
Better to put "legacy" on constants to avoid typos

Reply:
words of wisdom